### PR TITLE
Allow using vectors in lake

### PR DESCRIPTION
**Pull request by jamii**
When ZED_USE_VECTOR is set, assume that all vng versions of zng files are present, and transform PoolScan into VecLister+VecScan.

This also finally connects the demand inference to the vector load, but it turns out that the cost of loading data is tiny compared to the overhead of decoding/encoding, so even on queries where demand inference can remove almost all of the data loading we're still slower than just running the same query against zng.

### CODE REVIEW

File: lake/commits/snapshot.go [hunk 1]
```diff
@@ -3,6 +3,7 @@ package commits
 import (
 	"errors"
 	"fmt"
+	"golang.org/x/exp/maps"
```
Comment (nwt) [inline lake/commits/snapshot.go:6]: We're on Go 1.21 now.
```suggestion
	"maps"
```

Comment (jamii) [reply lake/commits/snapshot.go]: Huh, the maps packages doesn't have the Keys function :S 

I'll just rewrite that section.

File: vng/vector/primitive.go [hunk 2]
```diff
@@ -213,9 +212,10 @@ func (p *PrimitiveReader) ReadBytes() (zcode.Bytes, error) {
 func (p *PrimitiveReader) next() error {
 	segment := p.segmap[0]
 	p.segmap = p.segmap[1:]
-	if segment.Length > 2*MaxSegmentThresh {
-		return errors.New("corrupt VNG: segment too big")
-	}
+	//TODO Segments are currently disabled.
+	//if segment.Length > 2*MaxSegmentThresh {
+	//    return errors.New("corrupt VNG: segment too big")
+	//}
```
Comment (nwt) [inline vng/vector/primitive.go:218]: I don't see anything failing without the changes in this file. Unless I'm missing something, let's leave it as is.

File: compiler/optimizer/demand.go [hunk 3]
```diff
@@ -29,73 +42,87 @@ func inferDemandSeqOutWith(demands map[dag.Op]demand.Demand, demandSeqOut demand
...
+		return demand.None()
+	}
+	if expr == nil {
+		return demand.None()
 	}
+	var demandIn demand.Demand
 	switch expr := expr.(type) {
+	case *dag.Agg:
+		// Since we don't know how the expr.Name will transform the inputs, we have to assume demand.All.
+		demandIn = demand.Union(
```
Comment (nwt) [inline compiler/optimizer/demand.go:85]:
```suggestion
		return demand.Union(
```

File: vng/writer.go [hunk 4]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 const (
-	MaxSegmentThresh = vector.MaxSegmentThresh
+	MaxSegmentThresh = math.MaxInt
```
Comment (nwt) [inline vng/writer.go:15]: Should we make this change to vector.MaxSegmentThresh intead?

File: vector/vector_test.go [hunk 5]
```diff
@@ -41,3 +49,88 @@ func FuzzQuery(f *testing.F) {
 		fuzz.CompareValues(t, resultZNG, resultVNG)
 	})
 }
+
+const N = 10000000
+
+func BenchmarkReadZng(b *testing.B) {
```
Comment (nwt) [inline vector/vector_test.go:55]: This doesn't really have anything to do with using vectors in the lake so if you want to add it let's do so in a separate PR.

File: vng/vector/primitive.go (hunk 2 above)
Comment (jamii) [reply vng/vector/primitive.go:218]: On main:

```
> zq -f vng -o ints.vng 'yield this, this' ints.zng                                                      

> zq 'count()' ints.vng                                                                                  
ints.vng: corrupt VNG: segment too big
```

You changed the write path to make one big segment, but the read path still panics if the segment is over the max length.

File: vng/writer.go (hunk 4 above)
Comment (jamii) [reply vng/writer.go]: Good point.

**Review by nwt — COMMENTED** [review]
I was thinking we'd do something more along the lines of https://github.com/brimdata/zed/tree/noah-lake-vector, which is a little simpler and also more complete in that it falls back to ZNG if VNG isn't available.

**Review by nwt — APPROVED** [review]
Let's merge this and worry about simplification in another PR.
